SignalServiceKit/src/Util/OWSBackgroundTask.m

```diff
@@ -4,14 +4,290 @@
+        // b2) or there _was_ an active instance recently.
+        BOOL shouldHaveBackgroundTask = (!self.isAppActive && (self.expirationMap.count > 0 || self.continuityTimer));
+        BOOL hasBackgroundTask = self.backgroundTaskId != UIBackgroundTaskInvalid;
+
+        if (shouldHaveBackgroundTask == hasBackgroundTask) {
+            // Current state is correct.
+            return YES;
+        } else if (shouldHaveBackgroundTask) {
+            DDLogInfo(@"%@ Starting background task.", self.logTag);
+            return [self startBackgroundTask];
+        } else {
+            // Need to end background task.
+            DDLogInfo(@"%@ Ending background task.", self.logTag);
+            UIBackgroundTaskIdentifier backgroundTaskId = self.backgroundTaskId;
+            self.backgroundTaskId = UIBackgroundTaskInvalid;
+            [CurrentAppContext() endBackgroundTask:backgroundTaskId];
+            return YES;
+        }
+    }
+}
+
+// Returns NO if the background task cannot be begun.
+- (BOOL)startBackgroundTask
+{
+    OWSAssert(CurrentAppContext().isMainApp);
+
+    @synchronized(self)
+    {
+        OWSAssert(self.backgroundTaskId == UIBackgroundTaskInvalid);
+
+        self.backgroundTaskId = [CurrentAppContext() beginBackgroundTaskWithExpirationHandler:^{
+            // Supposedly [UIApplication beginBackgroundTaskWithExpirationHandler]'s handler
+            // will always be called on the main thread, but in practice we've observed
+            // otherwise.
+            //
+            // See:
+            // https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtaskwithexpiratio)
+            OWSAssert([NSThread isMainThread]);
+
+            [self backgroundTaskExpired];
+        }];
+
+        // If a background task could not be begun, call the completion block.
+        if (self.backgroundTaskId == UIBackgroundTaskInvalid) {
+            DDLogError(@"%@ background task could not be started.", self.logTag);
+
+            return NO;
+        }
+        return YES;
+    }
+}
+
+- (void)backgroundTaskExpired
+{
+    UIBackgroundTaskIdentifier backgroundTaskId;
+    NSDictionary<OWSTaskId, BackgroundTaskExpirationBlock> *expirationMap;
+
+    @synchronized(self)
+    {
+        backgroundTaskId = self.backgroundTaskId;
+        self.backgroundTaskId = UIBackgroundTaskInvalid;
+
+        expirationMap = [self.expirationMap copy];
+        [self.expirationMap removeAllObjects];
+    }
+
+    // It'd be nice to do this work synchronously, but it seems unsafe to
+    // depend on all expiration blocks being cheap.
+    //
+    // OWSBackgroundTask's API guarantees that completionBlock will always
+    // be called on the main thread, so facilitate that by dispatching async
+    // to main queue here.  That way we can ensure that we don't end the
+    // background task until all of the completion blocks have completed.
+    dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Dispatching async from the expiration handler seems like a bad idea

https://developer.apple.com/documentation/uikit/uiapplication/1623051-beginbackgroundtaskwithname?language=objc
> The handler is called synchronously on the main thread, blocking the app’s suspension momentarily while the app is notified.

Presumably the calling iOS code looks something like
```
-(void) taskDidExpire
{
    NSAssert([NSThread isMainThread]);
    callTheirExpirationHandler();
    terminateApp();
}
```

I read that as "whatever is blocking your app's suspension will run before any dispatched code." - so I'd say these blocks should be run in line.

Reply:
Done.